spec.md

```diff
@@ -0,0 +1,246 @@
+OctalLiteral   = "0o" [ "_" ] octal_digit   { [ "_" ] octal_digit   } [ "_" ] IntegerType .
+BinaryLiteral  = "0b" [ "_" ] ( "0" | "1" ) { [ "_" ] ( "0" | "1" ) } [ "_" ] IntegerType .
+
+IntegerType = TODO
+```
+
+where `IntegerType` is a suffix to specify the type of the value. Note that decimal literals cannot
+contain a leading `0` as part of a larger literal. This removes any possible ambiguity surrounding
+C-era octal literals, which are declared with leading zeroes.
+
+Negation is not specified here, as it not strictly part of the literal because it results from the
+unary negation operator: `-`.
+
+All of the standard integer types are available, so:
+
+#### Floating-point literals
+
+Floating-point literals are as expected. Unlike integer literals, there is only one kind:
+
+```
+FloatLiteral = ( "0" | { "0" … "9" } ) [ "_" ]
+               [ "." ( "0" … "9" ) { "_" | "0" … "9" } ]
+               [ ( "e" | "E" ) [ "+" | "-" ] ( "1" … "9" ) { "0" … "9" } ]
+```
+
+TODO: Explain this
+
+#### String literals
+
+There are both "normal" string literals and raw string literals. These both function as would
+be expected in many other languages, and evaluate at compile-time to a `&'static str`.
+
+```
+StringLiteral    = <"> { unicode_char | ByteEscape | UnicodeEscape | CharEscape } <"> .
+RawStringLiteral = "`" { unicode_char } "`" .
+```
+
+where escapes are any one of *byte escapes* (`ByteEscape`), *unicode escapes* (`UnicodeEscape`),
+or *character escapes* (`CharEscape`) - of which a handlful exist.
+
+```
+ByteEscape    = "\x" hex_digit hex_digit .
+UnicodeEscape = "\u{" { hex_digit } "}" .
+CharEscape    = "\" ( "n" | "r" | "t" | "\" | "0" | <"> | "'" ) .
+```
+
+Note that unicdode escapes may have **at most** 6 digits, representing a code point up to 24 bytes.
+A table has been placed below to give meaning to the character escapes.
+
+| Escape | Byte-value | Meaning          |
+|--------|------------|------------------|
+| `\n`   | 0x0A       | Newline          |
+| `\r`   | 0x0D       | Carriage-return  |
+| `\t`   | 0x09       | Tab              |
+| `\\`   | 0x5C       | Backslash        |
+| `\0`   | 0x00       | Null / Zero byte |
+| `\"`   | 0x22       | Double-quote     |
+| `\'`   | 0x27       | Single-quote     |
+
+A case for removing the vertical tab:
+[https://prog21.dadgum.com/76.html](https://prog21.dadgum.com/76.html)
+
+#### Character literals
+
+A [character](#characters) constitutes a single unicode scalar value, and so character literals
+allow values within those ranges. All of the escapes available from
+[string literals](#string-literals) are available here.
+
+```
+CharacterLiteral = "'" ( unicode_char | ByteEscape | UnicodeEscape | CharEscape ) "'" .
+```
+
+Character literals evaluate to a `char`.
+
+### Fn Declaration
+A fn declaration is used to define a function, for example:
+```
+  fn f(x) => ...  (where ... is an `expression`)
```

Comment:
I think there are very few low-level languages that allow function definitions to infer their types. Is there any benefit from doing this? It seems like a no-go for public functions, is there a use case for private ones?

Reply:
I think that makes sense for public functions, since their interface matters so making a change to it should be explicit. Plus, the inferred types might be looser then intended, in a way in which the code will works, but the type doesn't tell you about the intention of the function.
For private functions, I think the same logic applies, however I think it's neater if we allow a lambda functions type to be inferred?
Public variables should be explicitly typed in the low level too, which would mean that a public lambda function would need to have its types specified.